src/java.base/share/classes/java/util/zip/GZIPHeaderBuilder.java

```diff
@@ -0,0 +1,361 @@
+     *
+     * @since 18
+     */
+    public byte[] generateBytes(byte cm,
+                                byte[] extraFieldBytes,
+                                String filename,
+                                String fileComment) throws IOException {
+        CRC32 crc = new CRC32();
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+        // the head of header.
+        byte [] head = new byte[] {
+                (byte) GZIPHeaderData.GZIP_MAGIC,        // Magic number (short)
+                (byte)(GZIPHeaderData.GZIP_MAGIC >> 8),  // Magic number (short)
+                cm,                       // Compression method (CM)
+                flags,                    // Flags (FLG)
+                0,                        // Modification time MTIME (int)
+                0,                        // Modification time MTIME (int)
+                0,                        // Modification time MTIME (int)
+                0,                        // Modification time MTIME (int)
+                0,                        // Extra flags (XFLG)
+                GZIPHeaderData.OS_UNKNOWN                // Operating system (OS)
+        };
+        // write head.
+        baos.write(head);
+
+        if ((flags & GZIPHeaderData.FHCRC) == GZIPHeaderData.FHCRC) {
+            crc.update(head, 0, head.length);
+        }
+
+        // write extra field.
+        if ((flags & GZIPHeaderData.FEXTRA) == GZIPHeaderData.FEXTRA) {
+            /* extra field, per RFC-1952:
+             *     +---+---+=================================+
+             *     | XLEN  |...XLEN bytes of "extra field"...|
+             *     +---+---+=================================+
+             */
+            int xlen = extraFieldBytes.length;
+            if (xlen > 0xffff) {
+                throw new ZipException("extra field size out of range");
+            }
+            // write XLEN.
+            baos.write((byte)(xlen & 0xff));
+            baos.write((byte)((xlen >> 8) & 0xff));
+            /*
+             * Here do verification before write data.
+             */
+            if (!verifyExtraFieldLayout(extraFieldBytes)) {
+                throw new ZipException("incorrect extra field format");
+            };
+            baos.write(extraFieldBytes);
+            if ((flags & GZIPHeaderData.FHCRC) == GZIPHeaderData.FHCRC) {
+                crc.update((byte)(xlen & 0xff));
+                crc.update(((byte)(xlen >> 8) & 0xff));
+                crc.update(extraFieldBytes, 0, extraFieldBytes.length);
+            }
+        }
+        // write file name.
+        if ((flags & GZIPHeaderData.FNAME) == GZIPHeaderData.FNAME) {
+            /*
+             *    +=========================================+
+             *    |...original file name, zero-terminated...|
+             *    +=========================================+
+             */
+            byte[] filenameBytes = filename.getBytes("ISO-8859-1");
```

Comment:
Should use [`StandardCharsets.ISO_8859_1`](https://github.com/openjdk/jdk/blob/8876eae42993d4425ba9886dde94b08f6101a257/src/java.base/share/classes/java/nio/charset/StandardCharsets.java#L55) than string names. Using the charset object is significantly faster than looking up charsets by string names, see https://bugs.openjdk.java.net/browse/JDK-8272120